File: modules/schemas/chunithm.py
from datetime import date
from pydantic import BaseModel, ConfigDict
from pydantic.generics import GenericModel
from typing import Optional, List, Union, TypeVar, Generic

T = TypeVar("T")


class ResponseWithData(GenericModel, Generic[T]):
    status: Optional[int] = 200
    message: Optional[str] = "success"
    data: Optional[T] = None


class MusicInfoSchema(BaseModel):
    music_id: int
    title: str
    artist: str
    category: str
    version: Optional[str] = None
    releaseDate: Optional[date] = None
    isDeleted: bool
    deletedVersion: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class MusicDifficultySchema(BaseModel):
    music_id: int
    version: str
    diff0_const: Optional[float] = None
    diff1_const: Optional[float] = None
    diff2_const: Optional[float] = None
    diff3_const: Optional[float] = None
    diff4_const: Optional[float] = None

    model_config = ConfigDict(from_attributes=True)


class ChartDataSchema(BaseModel):
    difficulty: int
    creator: Optional[str] = None
    bpm: Optional[float] = None
    tap_count: Optional[int] = None
    hold_count: Optional[int] = None
    slide_count: Optional[int] = None
    air_count: Optional[int] = None
    flick_count: Optional[int] = None
    total_count: Optional[int] = None

    model_config = ConfigDict(from_attributes=True)


class DefaultServerSchema(BaseModel):
    im_id: str
    platform: str
    server: str

    model_config = ConfigDict(from_attributes=True)


class MusicBatchItemSchema(BaseModel):
    version: Optional[str] = None
    difficulty: List[Optional[float]]
    info: MusicInfoSchema


class MusicBatchResultSchema(BaseModel):
    __root__: dict[int, MusicBatchItemSchema]


class MusicBatchRequestSchema(BaseModel):
    music_ids: List[int]
    version: str


class MusicAliasSchema(BaseModel):
    id: Optional[int] = None
    alias: str


class BindingSchema(BaseModel):
    im_id: str
    platform: str
    server: Optional[str] = None
    aime_id: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class MusicDifficultyResponse(ResponseWithData[MusicDifficultySchema]):
    def __init__(self, **data):
        super().__init__(**data)


class MusicInfoResponse(ResponseWithData[MusicInfoSchema]):
    def __init__(self, **data):
        super().__init__(**data)


class AllMusicResponse(ResponseWithData[List[MusicInfoSchema]]):
    def __init__(self, **data):
        super().__init__(**data)


class ChartDataResponse(ResponseWithData[ChartDataSchema]):
    def __init__(self, **data):
        super().__init__(**data)


class DefaultServerResponse(ResponseWithData[DefaultServerSchema]):
    def __init__(self, **data):
        super().__init__(**data)


class BindingResponse(ResponseWithData[BindingSchema]):
    def __init__(self, **data):
        super().__init__(**data)


class AliasToMusicIDResponse(ResponseWithData[List[int]]):
    def __init__(self, **data):
        super().__init__(**data)


class AllAliasesResponse(ResponseWithData[List[str]]):
    def __init__(self, **data):
        super().__init__(**data)


class AddAliasResponse(ResponseWithData[MusicAliasSchema]):
    def __init__(self, **data):
        super().__init__(**data)

File: utils/types/chunithm.go
package types

import "time"

// ================= Chunithm Music Types =================

type ChunithmMusicInfo struct {
	MusicID        int        `json:"music_id"`
	Title          string     `json:"title"`
	Artist         string     `json:"artist"`
	Category       *string    `json:"category,omitempty"`
	Version        *string    `json:"version,omitempty"`
	ReleaseDate    *time.Time `json:"release_date,omitempty"`
	IsDeleted      *bool      `json:"is_deleted,omitempty"`
	DeletedVersion *string    `json:"deleted_version,omitempty"`
}

type ChunithmMusicDifficulty struct {
	MusicID int      `json:"music_id"`
	Version string   `json:"version"`
	Diff0   *float64 `json:"diff0_const,omitempty"`
	Diff1   *float64 `json:"diff1_const,omitempty"`
	Diff2   *float64 `json:"diff2_const,omitempty"`
	Diff3   *float64 `json:"diff3_const,omitempty"`
	Diff4   *float64 `json:"diff4_const,omitempty"`
}

type ChunithmChartData struct {
	Difficulty int      `json:"difficulty"`
	Creator    *string  `json:"creator,omitempty"`
	BPM        *float64 `json:"bpm,omitempty"`
	TapCount   *int     `json:"tap_count,omitempty"`
	HoldCount  *int     `json:"hold_count,omitempty"`
	SlideCount *int     `json:"slide_count,omitempty"`
	AirCount   *int     `json:"air_count,omitempty"`
	FlickCount *int     `json:"flick_count,omitempty"`
	TotalCount *int     `json:"total_count,omitempty"`
}

type ChunithmMusicBatchItem struct {
	Version    *string           `json:"version,omitempty"`
	Difficulty []*float64        `json:"difficulty"`
	Info       ChunithmMusicInfo `json:"info"`
}

// ================= Chunithm Binding Types =================

type ChunithmDefaultServer struct {
	UserID int    `json:"user_id"`
	Server string `json:"server"`
}

type ChunithmBinding struct {
	UserID int     `json:"user_id"`
	Server *string `json:"server,omitempty"`
	AimeID *string `json:"aime_id,omitempty"`
}

// ================= Chunithm Alias Types =================

type ChunithmMusicAlias struct {
	ID    int64  `json:"id,omitempty"`
	Alias string `json:"alias"`
}
